### DISCUSSION TITLE
Problem using Git basic commands - push

### DISCUSSION THREAD

**Discussion by ghost** [community/community#22054] [New to GitHub]
<p>Hello everyone, my name is Billy and am a student studying BSc. business intelligence. I have just started an online course about Front-End Web Development. First week course is about setting up Git and using basic Git commands. As you see from the screenshot, I commited four changes and would like to push these changes to my remote repository. However, this command is rejected. The error message hints to use “git pull” and I tried, terminal gave me another error:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/2d9ba1e4b6312d2786ee8dd204885de2b8f55801.png" data-download-href="/uploads/short-url/6vsVtxQiAMm9qqRiThA877rFCcV.png?dl=1" title="Screenshot from 2020-08-01 11-25-58" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/66633639/181068712-7d1218f9-aa86-4c42-a665-b3ed0ed0ba4d.png" alt="Screenshot from 2020-08-01 11-25-58" data-base62-sha1="6vsVtxQiAMm9qqRiThA877rFCcV" width="690" height="424" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/2/2d9ba1e4b6312d2786ee8dd204885de2b8f55801_2_690x424.png, https://aws1.discourse-cdn.com/github/original/2X/2/2d9ba1e4b6312d2786ee8dd204885de2b8f55801.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/2/2d9ba1e4b6312d2786ee8dd204885de2b8f55801.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/2d9ba1e4b6312d2786ee8dd204885de2b8f55801_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot from 2020-08-01 11-25-58</span><span class="informations">977×601 109 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Does anybody knows how I can solve this problem, so that the changes can be pushed to the remote repository? Thanks in advance!</p>

**Comment by airtower-luna** [comment]
<p>The “refusing to merge unrelated histories” message means that your local and remote repositories have absolutely no commits in common. That can mean a few things:</p>
<ol>
<li>You have the wrong remote repository (seems unlikely if it’s already configured as <code>origin</code>).</li>
<li>You edited the commit history back to the first commit (<code>git rebase</code>, <code>git commit --amend</code>, or maybe <code>git filter-branch</code>), so even if the content is the same the commits are all different.</li>
<li>It could also be that the remote repository was initialized with a commit (e.g. a default readme), but you don’t actually care about that.</li>
</ol>
<p>I’ll assume the second case here, but the third one can be handled the same way.</p>
<ul>
<li>Do <code>git fetch origin</code> to have the current state of the remote repository available to compare.</li>
<li>Check the commit logs of both the local and remote branches (<code>git log master</code>, <code>git log origin/master</code>).</li>
<li>You can also use <code>git diff origin/master..master</code> to see what your local branch changes relative to the remote one.</li>
</ul>
<p>Decide what you want to keep. That can be just one version (I’d guess the local one). If you also want to keep a few commits from remote you can use <code>git cherry-pick</code> to add them to the local branch. You could also create a new branch based on whichever branch most fits what you want to keep and then cherry pick, to have a sort of sandbox where you won’t destroy anything.</p>
<p>Once you have a branch that you are <em>absolutely sure</em> contains what you want to keep, you can force-push (<code>git push --force</code>) it to the remote repository. You have to be absolutely sure, because this will DESTROY EVERYTHING on the remote branch. :warning:</p>
<p>Theoretically you can also tell git to merge despite the unconnected histories, but that is almost always the wrong choice (which is why git refuses do to it by default).</p>

**Comment by ghost** [comment]
<p>Thank you very much for your feedback. I checked the commit logs of my local and my remote branches. I would like to keep my local one, that also means, I would like to push the local one to the remote repository. Do I need to tell the terminal which version (in my case the local one) I will keep? Because when I did git push --force right aright, it shows the following error:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/a32655e1e75d5d67fb0fc5b76865942761a7fde9.png" data-download-href="/uploads/short-url/nhhXAl4QtdqyLFhnH70wmR6n4hP.png?dl=1" title="Screenshot from 2020-08-01 12-27-48" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/66633639/181068717-f702bdee-09e5-4f04-8963-0d8e9fc796f4.png" alt="Screenshot from 2020-08-01 12-27-48" data-base62-sha1="nhhXAl4QtdqyLFhnH70wmR6n4hP" width="549" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/a/a32655e1e75d5d67fb0fc5b76865942761a7fde9_2_549x500.png, https://aws1.discourse-cdn.com/github/original/2X/a/a32655e1e75d5d67fb0fc5b76865942761a7fde9.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/a/a32655e1e75d5d67fb0fc5b76865942761a7fde9.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/a32655e1e75d5d67fb0fc5b76865942761a7fde9_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot from 2020-08-01 12-27-48</span><span class="informations">667×607 87.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by airtower-luna** [comment, marked as answer]
<p>Each branch in a git repository can have a remote “upstream” branch (also called “tracking” branch). If you do a <code>git pull</code> or <code>git push</code> (force or not) without further parameters it will pull from or push to there. The error message says that your local master branch doesn’t have a configured upstream branch.</p>
<p>If you want to set <code>origin/master</code> as the upstream branch for your local <code>master</code>, you can just use the command suggested by the error message, and add <code>--force</code> for the force-push:</p>
<pre><code class="lang-auto">git push --force --set-upstream origin master
</code></pre>
<p>If you don’t want to set the upstream branch leave out the <code>--set-upstream</code> option.</p>

**Comment by ghost** [comment]
<p>Thanks for the explanation! I tried the git push --force --set-upstream origin master command and problem was solved immediately. Now all commits in my local repository was synchronized to my remote repository. Honestly I don’t completely understand how it worked (and did not work)… but still thank you so much!</p>
